ClearCoat for MeshPhysicalMaterial

bhouston opened the pull request:
this is a relatively quick implementation of clearCoat for the Physical material.

There are two parameters introduced to the Physical material:
- clearCoat <float> which represents the effective thickness of the clear coat.  0 is no clear coat, 1 is full clear coat.
- clearCoatRoughness <float> which represents the GGX roughness of the clear coat.

I have also added an example of the clear coat in action as:

examples/webgl_materials_variations_physical.html

This clear coat implementation using a GGX-Smith- Correlated visibility function.  I understand that others use the Kelemen visibility function but it is quite hard to tell the difference.  We can upgrade to the Kelemen visibility function in the future and it will have very little visible difference in the results.

It looks like this:

![image](https://cloud.githubusercontent.com/assets/588541/15652225/b1f1abc8-2652-11e6-8cbe-08a5706dd1c0.png)

Please note that I am using the UE4 parameterization of clearCoat and clearCoatRoughness rather than the Disney parameters of clearCoat and clearCoatGlossiness.

/ping @WestLangley

bhouston commented:
BTW the main reason I didn't implement a Kelemen geometry function is that the indirect specular GGX Environment BDRF is opaque, I can not figure out how to modify it to also use a Kelement geometry function instead of the GGX geometry function to keep it in sync with the direct specular lighting.  This implementation is at least consistent between direct and indirect specular even though it is still reliant upon the GGX geometry function.

Here is a Kelemen visibility function (cut-and-pasted from Clara.io's code base) as we could use:

```
float G_Kelemen( float dotHV ) {
  return 1.0 / ( 4.0 * pow2( dotHV ) + EPSILON );
}
```

WestLangley commented on src/renderers/shaders/ShaderLib.js:
```diff
@@ -93,7 +93,9 @@ THREE.ShaderLib = {
-				"envMapIntensity" : { type: "1f", value: 1 } // temporary
+				"envMapIntensity" : { type: "1f", value: 1 }, // temporary
+				"clearCoat": { type: "1f", value: 0 },
+				"clearCoatRoughness": { type: "1f", value: 0 }
```
These new properties are not properties of `StandardMaterial`, and belong in `THREE.ShaderLib[ 'physical' ]`. In fact, there is already a "future" placeholder for them there. :)

WestLangley commented:
@bhouston Thanks! I have implemented the same in my fork, and also used GGX. I do not think the clear coat model matters that much.

I will compare our approaches as soon as I can, and see if there are any modeling differences that appear significant.

I used the uncapitalized form `.clearcoat` / `.clearcoatRoughness` -- personal preference, of course.

bhouston commented:
@WestLangley thanks for the partial review.  I've moved the uniforms into the right place. :)  I think I'll keep "clearCoat" like that because it is generally used as two words in most references I have found: "clear coat".  Good to know that you went with GGX as well.

I believe that the use of the Kelemen visible/geometry distribution for clear coats originates from this paper (by Kelemen of course):

http://sirkan.iit.bme.hu/~szirmay/scook.pdf

WestLangley commented:
@bhouston All the pieces are in place and I like your refactoring. But your model is losing too much energy when `clearCoat` is increased. I have a proposed fix though.

I should preface this by saying there are no right answers, and we may each choose to make different compromises. We just have to do something that is reasonable.

Here are results of my proposed fix. The underlying material in this example is rough with a (maximal) _super-smooth_ clear coat.

 
![screen shot 2016-05-30 at 4 38 45 pm](https://cloud.githubusercontent.com/assets/1000017/15658437/a6b76346-268a-11e6-8a0f-c9e3de2e2676.png)


And the same underlying material with a maximal _super-rough_ clear coat.

![screen shot 2016-05-30 at 4 39 02 pm](https://cloud.githubusercontent.com/assets/1000017/15658446/b374e806-268a-11e6-88c3-a9cb0f8231bb.png)

These compare to the same two cases -- smooth clear coat and rough clear coat -- using the model in this PR. Notice how energy is lost from the underlying material.

![screen shot 2016-05-30 at 4 42 04 pm](https://cloud.githubusercontent.com/assets/1000017/15658462/c0e12c52-268a-11e6-8618-6a2d6a4e65d8.png)

![screen shot 2016-05-30 at 4 42 19 pm](https://cloud.githubusercontent.com/assets/1000017/15658471/e186f8f6-268a-11e6-8089-d97155fef8b3.png)


In the model I implemented in my fork, I made the following assumptions.

1. The clear coat does not reflect any light diffusely. It is transparent. That energy is passed through to the underlying material, where it is absorbed, or reflected.
2. The specular energy reflected by the clear coat is minimal. For that reason, it is OK -- for now anyway -- to ignore the energy conservation calculation that would otherwise be required.

Here are the suggested changes to your model. Posted for clarity, not efficiency.

```glsl
void RE_Direct_Physical( const in IncidentLight directLight, const in GeometricContext geometry, const in PhysicalMaterial material, inout ReflectedLight reflectedLight ) {

	float dotNL = saturate( dot( geometry.normal, directLight.direction ) );

	vec3 irradiance = dotNL * directLight.color;

	#ifndef PHYSICALLY_CORRECT_LIGHTS

		irradiance *= PI; // punctual light

	#endif

	vec3 specularRadiance = irradiance * BRDF_Specular_GGX( directLight, geometry, material.specularColor, material.specularRoughness );
	vec3 diffuseRadiance = irradiance * BRDF_Diffuse_Lambert( material.diffuseColor );

	#ifndef STANDARD

		vec3 clearCoatSpecularRadiance = irradiance * BRDF_Specular_GGX( directLight, geometry, vec3( material.clearCoat ), material.clearCoatRoughness );

		//specularRadiance = mix( specularRadiance, clearCoatSpecularRadiance, material.clearCoat );
		//diffuseRadiance = mix( diffuseRadiance, vec3( 0.0 ), material.clearCoat );
		specularRadiance += material.clearCoat * clearCoatSpecularRadiance;

	#endif

	reflectedLight.directSpecular += specularRadiance;
	reflectedLight.directDiffuse += diffuseRadiance;

}

void RE_IndirectDiffuse_Physical( const in vec3 irradiance, const in GeometricContext geometry, const in PhysicalMaterial material, inout ReflectedLight reflectedLight ) {

	vec3 diffuseRadiance = irradiance * BRDF_Diffuse_Lambert( material.diffuseColor );

	//#ifndef STANDARD

	//	diffuseRadiance = mix( diffuseRadiance, vec3( 0.0 ), material.clearCoat );

	//#endif

	reflectedLight.indirectDiffuse += diffuseRadiance;

}

void RE_IndirectSpecular_Physical( const in vec3 radiance, const in vec3 clearCoatRadiance, const in GeometricContext geometry, const in PhysicalMaterial material, inout ReflectedLight reflectedLight ) {

	vec3 specularRadiance = radiance * BRDF_Specular_GGX_Environment( geometry, material.specularColor, material.specularRoughness );

	#ifndef STANDARD

		vec3 clearCoatSpecularRadiance = clearCoatRadiance * BRDF_Specular_GGX_Environment( geometry, vec3( material.clearCoat * DEFAULT_SPECULAR_COEFFICIENT ), material.clearCoatRoughness );

		//specularRadiance = mix( specularRadiance, clearCoatSpecularRadiance, material.clearCoat );
		specularRadiance += material.clearCoat * clearCoatSpecularRadiance;

	#endif

	reflectedLight.indirectSpecular += specularRadiance;

}
```

I am open to other changes, such a modeling the clear coat as a polyurethane. I may still have a few minor additional suggestions later -- after some experimentation.

And of course, thank you for this PR!

bhouston commented:
I followed the energy conservation model because I was following the UE4 clear coat model -- this is actually how they do it.  But you seem to be suggesting what Burley suggests in the Disney PBR paper:

> For our clearcoat layer, we use a fixed ior of 1.5, representative of polyurethane, and instead
allow artists to scale the overall strength of the layer using the clearcoat parameter. The normalized
parameter range corresponds to an overall scale of [0, 0.25]. This layer, even though it has a large
visual impact, represents a relatively small amount of energy so we don’t subtract any energy from the base layer. When set to zero, the clearcoat layer is effectively disabled and incurs no cost.

Quoted from pg 16 of https://disney-animation.s3.amazonaws.com/library/s2012_pbs_disney_brdf_notes_v2.pdf

Disney model, like you suggest, does not engage in energy conservation, but they also limit the strength of the clear coat to go to 0.25, as that seems to be the meaningful range (sort of like how we limit reflectance to the range of [0,0.16].  Thus when we set ```material.clearCoat``` in ```lights_physical_fragment.glsl``` we would do this:

```
material.clearCoat = saturate( clearCoat ) * 0.25;
```

And this would work well with the rest of your proposed changes.

What do you think?

WestLangley commented:
Yes. That must have been what I read, too. : - )

>I followed the energy conservation model 

Yes, but I do not think what you implemented is the correct energy conservation model. You would have to do an integral to see how much total energy is reflected specularly, and then subtract that from the energy that is available for the next layer. I expect the total energy reflected by the clear coat layer is very small (assuming no diffuse reflectance), but I haven't done the math to quantify.

I actually remapped `clearCoat` like so:
```glsl
material.clearCoat = 0.5 * clearCoat;
```
But I like it not remapped now. : - )  Can we leave it not remapped for now so we can experiment with it?

bhouston commented:
I just implemented what you asked and I do not like the results as it is missing the richness that the energy conserving model added:

![image](https://cloud.githubusercontent.com/assets/588541/15660176/a012b54c-26a6-11e6-907c-fb4155092d54.png)

I actually like the idea of not allowing for one to use only clear coatas I had it in the original PR (thus I like the idea of limiting it to 0.25 as Burley does as that seems reasonable) , but the energy conservation adds so much realism -- otherwise the artist has to do that themselves, and know that they have to do that (which is hard on artists to remember that they have to always do this compensation.)

My concern is that you are not familiar with the realism that energy conservation adds.

bhouston commented:
Energy conservation is actually a major part of PBR models and a major reason why they are realistic.  The energy conservation that metalness does to diffuse is the best example of how energy conservation adds tons of realism.

Now I am not saying I am doing energy conservation correct in this case - I maybe should be doing it based on the result of the Schlick fresnel term.

I think the simplest is given in equation (22) in this classic paper by Schlick on how one does simple energy conservation through layered materials:

https://www.cs.virginia.edu/~jdl/bib/appearance/analytic%20models/schlick94b.pdf

bhouston commented:
> I expect the total energy reflected by the clear coat layer is very small (assuming no diffuse reflectance), but I haven't done the math to quantify.

I believe if you calculate the Schlick Fresnel term, that is how much is reflected by the specular layer, and ( 1 - Schlick_Fresnel ) is what is available to the layer underneath.

We could do this for clearCoat as well as the standard specular lobe (well we would have to chain the two energy preservations together) and ThreeJS's physical model would be energy preserving.

WestLangley commented:
Can you remove the build from the PR?

>My concern is that you are not familiar with the realism that energy conservation adds.

Oh, I support your desire to handle energy conservation. But you are not implementing an energy conservation model, IMO. You are mixing materials. That is not the correct math.

I am arguing for now that since the excess energy is maybe 1%, we ignore it until we can figure out how to handle energy conservation correctly.

The clear coat is only reflecting specular energy, remember. It should look like a gloss or satin finish on top of the bass material.

bhouston commented:
> I am arguing for now that since the excess energy is maybe 1%, we ignore it until we can figure out how to handle energy conservation correctly.

At grazing angles the Fresnel term goes to 1.0 right?  Thus the clear coat specular goes to 1.0 at grazing angles and then is multiplied by 0.25.  And the underlying specular lobe will go to 1.0 at grazing angles as well.  Thus together we are at 1.25x incoming energy at grazing angles for a smooth surface based on specular alone (and that excludes diffuse contributions) if we use the Burley parameterization.  It would be at least 2x incoming energy based on specular + clear coat at grazing angles on a smooth surface before I switched to the Burley parameterization.

We are not energy perserving.

Please see Schlick's paper on energy preserving.  As long as we do not model internal reflections we can actually do layering really easily.

WestLangley commented:
Yes, as I said, the model I implemented is not energy-conserving. (I have no objection to adding energy-conservation, though.)

Let's have a go at adding in a reasonable energy-conserving term.

Can you remove the build from the PR?

bhouston commented:
> Can you remove the build from the PR?

I'll commit the current build in this PR on top, with the new PR collapsing that mrdoob is doing, it will have the effect of removing it from the three.js git repo history.

bhouston committed:
clear coat implementation.

bhouston committed:
add physical variations to examples.

bhouston commented:
I've removed any energy conservation because even though I think it is important, it will make this PR hard to get accepted.  Thus I've kept it simple and even adopted the 0.5 parameterization of @WestLangley rather then the 0.25 parameterization of Burley.

Energy conservation will be done as a separate future PR.

WestLangley commented:
>Energy conservation will be done as a separate future PR.

OK.

This PR is good-to-go as far as I am concerned.

bhouston committed:
add example file for physical materials

mrdoob commented:
Uh oh, seems like my `type:` clean up created conflicts. Do you mind pulling from `dev`?

bhouston committed:
adjust clearcoat range to 0-1 as original.

bhouston committed:
Merge branch 'dev' of git@github.com:mrdoob/three.js.git into clearcoat

# Conflicts:
#	src/renderers/shaders/ShaderLib.js

bhouston committed:
switch example color to gray from yellow.

bhouston commented:
I merged in dev, I changed the clearCoat range to 0-1 as @WestLangley suggested [here](https://github.com/mrdoob/three.js/pull/9038#issuecomment-222570158) and changed the example color to gray from yellow.  It should be good to go.

mrdoob commented:
Thanks!

WestLangley commented:
@bhouston 

Earlier, I wrote:

>You would have to do an integral to see how much total energy is reflected specularly, and then subtract that from the energy that is available for the next layer. I expect the total energy reflected by the clear coat layer is very small (assuming no diffuse reflectance), but I haven't done the math to quantify.

Directional-hemispherical reflectance, or directional albedo, is the fraction of radiant energy scattered. It is the integral of the cosine-weighted BRDF over all viewing directions.

	albedo( ωi ) = ∫ fr( ωo, ωi ) cos( θo ) dωo		for a given incoming direction ωi

In the above formula, `fr()` is the BRDF, and the integral is over the hemisphere defined by the surface normal.

Due to energy conservation, albedo ≤ 1.

For our clear coat GGX BRDF, this integral must be evaluated numerically. It can be parameterized by the clear coat roughness, and the incoming light direction. Here is the surface plot:

![directional albedo](https://cloud.githubusercontent.com/assets/1000017/16360081/9bd7c244-3b1c-11e6-8c5c-b2ba723b8a87.png)

The total energy reflected by the clear coat layer is quite small, except when the surface is smooth, and the incoming grazing angle is about 70 degrees or greater.

So we can modify our rendering equation to subtract this reflected energy from that which would normally be available to the base layer.

I have hacked up an example [here](https://dl.dropboxusercontent.com/u/63997063/Three/ClearCoatEnergyTest/Three.JS.ClearCoat.Energy.Testbed.html).

In the example, there is a checkbox to toggle off clear coat energy conservation. As you can see, the effect is minimal in most cases.

Note that in the example, I approximated the directional albedo (plotted above) with a simple function. In practice, we would have to save the directional albedo to a data texture, or use MATLAB to find a better functional form.

bhouston commented:
Sweet.  What equation did you derive in the end?  Others have suggested
using the fresnel which has gross similarity to what you described but I
suspect the details are fairly different.  Beautiful results, this is a
really nice contributions .

Best regards,
Ben Houston
http://Clara.io Online 3d modeling and rendering
On Jun 25, 2016 9:59 PM, "WestLangley" <notifications@github.com> wrote:

> @bhouston <https://github.com/bhouston>
>
> Earlier, I wrote:
>
> You would have to do an integral to see how much total energy is reflected
> specularly, and then subtract that from the energy that is available for
> the next layer. I expect the total energy reflected by the clear coat layer
> is very small (assuming no diffuse reflectance), but I haven't done the
> math to quantify.
>
> Directional-hemispherical reflectance, or directional albedo, is the
> fraction of radiant energy scattered. It is the integral of the
> cosine-weighted BRDF over all viewing directions.
>
> albedo( ωi ) = ∫ fr( ωo, ωi ) cos( θo ) dωo      for a given incoming direction ωi
>
> In the above formula, fr() is the BRDF, and the integral is over the
> hemisphere defined by the surface normal.
>
> Due to energy conservation, albedo ≤ 1.
>
> For our clear coat GGX BRDF, this integral must be evaluated numerically.
> It can be parameterized by the clear coat roughness, and the incoming light
> direction. Here is the surface plot:
>
> [image: directional albedo]
> <https://cloud.githubusercontent.com/assets/1000017/16360081/9bd7c244-3b1c-11e6-8c5c-b2ba723b8a87.png>
>
> The total energy reflected by the clear coat layer is quite small, except
> when the surface is smooth, and the incoming grazing angle is about 70
> degrees or greater.
>
> So we can modify our rendering equation to subtract this reflected energy
> from that which would normally be available to the base layer.
>
> I have hacked up an example here
> <https://dl.dropboxusercontent.com/u/63997063/Three/ClearCoatEnergyTest/Three.JS.ClearCoat.Energy.Testbed.html>
> .
>
> In the example, there is a checkbox to toggle off clear coat energy
> conservation. As you can see, the effect is minimal in most cases.
>
> Note that in the example, I approximated the directional albedo (plotted
> above) with a simple function. In practice, we would have to save the
> directional albedo to a data texture, or use MATLAB to find a better
> functional form.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mrdoob/three.js/pull/9038#issuecomment-228579949>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAj6_TVxA788M2cvPtdEW5lnssWstXQzks5qPd0FgaJpZM4Ip2iM>
> .
>

WestLangley commented:
@bhouston 
>What equation did you derive in the end?

Here is the approximation I used in the example. It is only meant to be used for our particular clear-coat layer. The approximation should be improved -- or replaced with a texture lookup.

```glsl
float clearCoatDHRApprox( const in float roughness, const in float dotNL ) {

    return DEFAULT_SPECULAR_COEFFICIENT + ( 1.0 - DEFAULT_SPECULAR_COEFFICIENT ) * ( pow( 1.0 - dotNL, 5.0 ) * pow( 1.0 - roughness, 2.0 ) );

}
```

The falloff when roughness increases is due to the shadow/masking effect.

bhouston commented:
@WestLangley wrote:

```
float clearCoatDHRApprox( const in float roughness, const in float dotNL ) {
    return DEFAULT_SPECULAR_COEFFICIENT + ( 1.0 - DEFAULT_SPECULAR_COEFFICIENT ) * ( pow( 1.0 - dotNL, 5.0 ) * pow( 1.0 - roughness, 2.0 ) );
}
```

I notice the above is the Schlick Fresnel equation modified by a ```pow2( 1.0 - roughness )```.  Thus I could rewrite it as this if we had a monochromatic Schlick function (which we should totally add):
```
float clearCoatDHRApprox( const in float roughness, const in float dotNL ) {
    return F_Schlick( DEFAULT_SPECULAR_COEFFICIENT ) * pow2( 1.0 - roughness );
}
```

I undertand where the Schlick Fresnel comes from, I suggested this as the energy conservation term in [this comment previously](https://github.com/mrdoob/three.js/pull/9038#issuecomment-222571658) as it is one of the standard methods to preserve this type of energy in layered materials.

What I do not understand is the ```pow2( 1.0 - roughness )``` term.  I do now know your derivation so I can not follow its introduction.

I actually am a bit confused by this addition it because it seems that the roughness affects the total energy reflected, which if I understand roughness correctly, it should not affect the total energy reflected, rather just in which direction it is reflected -- e.g. how it works when sampling pre-blurred environment maps.  But changes the directionality of reflected energy should not change the results of an integral summation, because the integral should sum across all directions -- thus it comes back to my confusion as to why roughness appears to be changing the total energy reflected.

I'd like to confirm the derivation of this energy conservation function.

WestLangley commented:
@bhouston The approximation is not "derived" -- it was defined to approximate the true directional hemispherical reflectance as pictured above. It was only used so I could show something quickly. We need to use the actual directional albedo -- or a better approximation -- in production.

>I actually am a bit confused by this addition it because it seems that the roughness affects the total energy reflected

As I noted above, it is the shadowing/masking effect that causes the directional hemispherical reflectance to be reduced as roughness is increased.

bhouston commented:
@WestLangley, okay I'm good with this, thanks for humoring me.

I'll make a PR that uses a Schlick Fresnel (per eqn 22 and 24 of [Schlick's 1994 paper](http://igorsklyar.com/system/documents/papers/28/Schlick94.pdf) ) combined with your self-shadowing term of ```pow2( 1 - roughness)``` for energy conservation.

WestLangley commented:
@bhouston I already have the PR ready-to-go if you are happy with the approximation.

bhouston commented:
@WestLangley, go for it!  Clara.io ha been using this Schlick Fresnel energy conservation model (without the self-shadowing) for the last year and it produces great results.

WestLangley commented:
@bhouston Even though what you are doing is not mathematically correct -- the only thing that is correct is

    albedo( ωi ) = ∫ fr( ωo, ωi ) cos( θo ) dωo 

-- our model does not handle 2nd bounces, so energy is lost. So the truth is somewhere in the middle.

The approach I have described, which is consistent with our model, will show less of an effect than your approach will. You may not like it if you are used to what you have implemented in-house. We can always change it later. Every implementation is a compromise. : - )
